pallets/stateful-storage/src/types.rs

```diff
@@ -0,0 +1,299 @@
+	/// Message Source Account identifier
+	#[codec(compact)]
+	pub msa_id: MessageSourceId,
+
+	/// Schema id of this storage
+	#[codec(compact)]
+	pub schema_id: SchemaId,
+
+	/// Page id of this storage
+	#[codec(compact)]
+	pub page_id: PageId,
+
+	/// Hash of targeted page to avoid race conditions
+	#[codec(compact)]
+	pub target_hash: PageHash,
+
+	/// The block number at which the signed proof will expire
+	pub expiration: T::BlockNumber,
+}
+
+/// A generic page of data which supports both Itemized and Paginated
+#[derive(Encode, Decode, TypeInfo, MaxEncodedLen, Debug, Default)]
+#[scale_info(skip_type_params(PageDataSize))]
+#[codec(mel_bound(PageDataSize: MaxEncodedLen))]
+pub struct Page<PageDataSize: Get<u32>> {
+	/// Data for the page
+	/// - Itemized is limited by [`Config::MaxItemizedPageSizeBytes`]
+	/// - Paginated is limited by [`Config::MaxPaginatedPageSizeBytes`]
+	pub data: BoundedVec<u8, PageDataSize>,
+}
+
+/// An internal struct which contains the parsed items in a page
+#[derive(Debug, PartialEq)]
+pub struct ParsedItemPage<'a> {
+	/// Page current size
+	pub page_size: usize,
+	/// A map of item index to a slice of blob (including header is optional)
+	pub items: BTreeMap<u16, &'a [u8]>,
+}
+
+impl<PageDataSize: Get<u32>> Page<PageDataSize> {
+	/// Check if the page is empty
+	pub fn is_empty(&self) -> bool {
+		self.data.is_empty()
+	}
+
+	/// Retrieve the hash of the page
+	pub fn get_hash(&self) -> PageHash {
+		if self.is_empty() {
+			return PageHash::default()
+		}
+		let mut hasher = XxHash64::with_seed(0);
+		self.hash(&mut hasher);
+		let value_bytes: [u8; 4] =
+			hasher.finish().to_be_bytes()[..4].try_into().expect("incorrect hash size");
+		PageHash::from_be_bytes(value_bytes)
+	}
+}
+
+/// PartialEq and Hash should be both derived or implemented manually based on clippy rules
+impl<PageDataSize: Get<u32>> Hash for Page<PageDataSize> {
+	fn hash<H: Hasher>(&self, state: &mut H) {
+		state.write(&self.data[..]);
+	}
+}
+
+/// PartialEq and Hash should be both derived or implemented manually based on clippy rules
+impl<PageDataSize: Get<u32>> PartialEq for Page<PageDataSize> {
+	fn eq(&self, other: &Self) -> bool {
+		self.data.eq(&other.data)
+	}
+}
+
+impl<PageDataSize: Get<u32>> From<BoundedVec<u8, PageDataSize>> for Page<PageDataSize> {
+	fn from(bounded: BoundedVec<u8, PageDataSize>) -> Self {
+		Self { data: bounded }
+	}
+}
+
+impl<PageDataSize: Get<u32>> TryFrom<Vec<u8>> for Page<PageDataSize> {
+	type Error = ();
+
+	fn try_from(data: Vec<u8>) -> Result<Self, Self::Error> {
+		let bounded: BoundedVec<u8, PageDataSize> = BoundedVec::try_from(data).map_err(|_| ())?;
+		Ok(Page::from(bounded))
+	}
+}
+
+impl<PageDataSize: Get<u32>> Page<PageDataSize> {
+	/// Applies all actions to specified page and returns the updated page
+	/// This has O(n) complexity when n is the number of all the bytes in that itemized storage
+	pub fn apply_item_actions<T: Config>(
```

Comment:
This is interesting. I do not see


pub type ItemizedPage<T> = Page<Config<Self = T>::MaxItemizedPageSizeBytes>

pub type PaginatedPage<T> = Page<Config<Self = T>::MaxPaginatedPageSizeBytes>


It seems that apply_item_actions is only used for Page<Config<Self = T>::MaxItemizedPageSizeBytes> however, it is only needed for ItemizedPage.

I think it might be easier to follow if we create a trait called something like
```
pub trait CommitActions {
 ...
}
```
And then we implemented on ItemizedPage.

By doing this we target only the type we want to implement that behavior.

Reply:
Created a trait to handle itemized ones